docs(logging): log4j2 support updates

PaulRMellor opened the pull request:
**Documentation**

Updates the logging content to reflect new support and Kafka 4.0 adoption of log4j2
- Removes/updates references to log4j1 and log4j1 logger names
- Updates logging content in the deploying guide 
  - Refreshes logging config overview
- streamlines logging information in API reference guide to remove repetition


### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

fvaleri commented on documentation/api/io.strimzi.api.kafka.model.connect.KafkaConnectSpec.adoc:
```diff
@@ -114,15 +107,13 @@ spec:
-      log4j.logger.org.apache.kafka.connect.runtime.WorkerSourceTask: TRACE
-      log4j.logger.org.apache.kafka.connect.runtime.WorkerSinkTask: DEBUG
+      rootLogger.level: INFO
+      logger.org.apache.kafka.connect.runtime.WorkerSourceTask: TRACE
+      logger.org.apache.kafka.connect.runtime.WorkerSinkTask: DEBUG
```
```suggestion
      logger.sourcetask.name = org.apache.kafka.connect.runtime.WorkerSourceTask
      logger.sourcetask.level = TRACE
      logger.sinktask.name = org.apache.kafka.connect.runtime.WorkerSinkTask
      logger.sinktask.level = DEBUG
```

fvaleri commented on documentation/api/io.strimzi.api.kafka.model.connect.KafkaConnectSpec.adoc:
```diff
@@ -80,8 +80,8 @@ In this case, fix the configuration so that the Cluster Operator can roll out th
 
-* `connect.root.logger.level`
-* `log4j.logger.org.reflections`
+* `rootLogger.level`
+* `logger.org.reflections`
```
```suggestion
* `rootLogger.*`
* `logger.reflections.name = org.reflections`
  `logger.reflections.*`
```

fvaleri commented on documentation/api/io.strimzi.api.kafka.model.kafka.KafkaClusterSpec.adoc:
```diff
@@ -119,30 +119,21 @@ Be aware that if the configured image is not compatible with Strimzi images, it
-
-Kafka uses the Apache `log4j` logger implementation.
-
+* `logger.kafka`
+* `logger.org.apache.kafka`
+* `logger.kafka.request.logger`
+* `logger.kafka.network.RequestChannel$`
+* `logger.org.apache.kafka.controller`
+* `logger.kafka.log.LogCleaner`
+* `logger.state.change.logger`
+* `logger.kafka.authorizer.logger`
```
I think this is not correct. See my previous comments and also the examples included in the following PR for each major component.

https://github.com/strimzi/strimzi-kafka-operator/pull/11176/files#diff-96dbf0bf20d7d83f0a33459b088dd6223e3c46f114c989e177c335505369d9f6

fvaleri commented on documentation/api/io.strimzi.api.kafka.model.kafka.KafkaClusterSpec.adoc:
```diff
@@ -154,18 +145,15 @@ spec:
-        log4j.logger.kafka.request.logger: DEBUG
-        log4j.logger.io.strimzi.kafka.oauth: DEBUG
-        log4j.logger.org.openpolicyagents.kafka.OpaAuthorizer: DEBUG
+        rootLogger.level: INFO
+        logger.kafka.coordinator.transaction: TRACE
+        logger.kafka.log.LogCleanerManager: DEBUG
+        logger.kafka.request.logger: DEBUG
+        logger.io.strimzi.kafka.oauth: DEBUG
```
Same as previous comments.

fvaleri commented on documentation/modules/operators/ref-operator-cluster-logging-configmap.adoc:
```diff
@@ -7,26 +7,32 @@
+* How often logging configuration is loaded
+
+Netty is a framework used in Strimzi for network communication.
+The `monitorInterval` setting determines how often in seconds the logging configuration is dynamically reloaded.
```
Should we say that `monitorInterval` is set to 30 seconds by default? This is different from log4j2 default value of 0 (i.e. reloading disabled).

PaulRMellor commented:
Thanks for the comments @fvaleri 
I've updated the examples as directed.
I've also updated the section on adding logging filters (appenders) based on your feedback.

PaulRMellor commented on documentation/api/io.strimzi.api.kafka.model.connect.KafkaConnectSpec.adoc:
```diff
@@ -80,8 +80,9 @@ In this case, fix the configuration so that the Cluster Operator can roll out th
 
-* `connect.root.logger.level`
-* `log4j.logger.org.reflections`
+* `rootLogger.*`
+* `logger.reflections.name = org.reflections`
+* `logger.reflections.*`
```
Thanks @scholzj. Before I proceed, I just want to confirm how you see the presentation. Like this?

> Kafka Connect has its own configurable loggers:
> 
> * `rootLogger`
> * `reflections`

As an alternative to adding the log4j1 examples back, we don't want to add a note in the relevant logging sections of the API reference to say something like: this is for Kafka 4/log4j2, for Kafka 3.9/log4j1 go look at 0.45 doc?

Assuming, with the addition of a few comments (as suggested), we can keep the deploying content as is

scholzj replied:
Yeah, I thikn just mentioning the Log4j1 and linking to 0.45.0 docs is fine.

PaulRMellor commented on documentation/api/io.strimzi.api.kafka.model.bridge.KafkaBridgeSpec.adoc:
```diff
@@ -37,7 +37,7 @@ You can replace `_<operation-id>_` in the `logger.__<operation-id>__` logger to
 You can change the log level on each endpoint to create fine-grained logging information about the incoming and outgoing HTTP requests.
 
-Each logger has to be configured assigning it a `name` as `http.openapi.operation.__<operation-id>__`.
+Each logger has to be configured assigning it a `name` as `http.openapi.operation.<operation_id>`.
```
I've refreshed this section. Made it more concise and clearer.

PaulRMellor commented on documentation/api/io.strimzi.api.kafka.model.connect.KafkaConnectSpec.adoc:
```diff
@@ -92,19 +93,12 @@ Use a curl request to get a complete list of Kafka Connect loggers running from
 
-Kafka Connect uses the Apache `log4j` logger implementation.
-
+Kafka Connect uses the Apache `log4j2` logger implementation.
```
I've added note, as discussed.

fvaleri commented on documentation/api/io.strimzi.api.kafka.model.bridge.KafkaBridgeSpec.adoc:
```diff
@@ -13,10 +13,10 @@ Configuration options relate to:
+* `bridge`
 
-You can replace `_<operation-id>_` in the `logger.__<operation-id>__` logger to set log levels for specific operations:
+You can also log levels for specific operations based on endpoints:
```
```suggestion
You can also set log levels for specific operations based on endpoints:
```

fvaleri reviewed (approved):
LGTM. Thanks.

scholzj commented on documentation/api/io.strimzi.api.kafka.model.bridge.KafkaBridgeSpec.adoc:
```diff
@@ -30,44 +30,38 @@ You can replace `_<operation-id>_` in the `logger.__<operation-id>__` logger to
-Each logger has to be configured assigning it a `name` as `http.openapi.operation.__<operation-id>__`.
-For example, configuring the logging level for the `send` operation logger means defining the following:
+[source,properties]
+----
+logger.healthy.name = http.openapi.operation.healthy
+logger.healthy.level = WARN
+logger.ready.name = http.openapi.operation.ready
+logger.ready.level = WARN
+----
```
This is still a bit confusing. If these loggers are defined, you should not need to create them.

@ppatierno Can you please clarify it? Do these loggers really exist? Or how is it? They do not seem to be in the Log4j2 configuration file.

PaulRMellor replied:
Thanks. I see them in strimzi-kafka-bridge/config
/log4j2.properties and strimzi-kafka-operator/cluster-operator/.../default-logging
/KafkaBridgeCluster.properties
Let's see what Paolo says

scholzj replied:
These two yes, that would mean you can leave out the `.name` lines. But not the other ones - they do not exist there.

scholzj replied:
That is getting the logger. Does it also define it? If so, does it define it only when the method is actually called? I'm not sure this realyl answers the question if the logger should be defined in the configuration or not (and just the log level can be configured).

ppatierno replied:
I am pretty sure I am going to say something you already well know but I still get confused on what you are trying to say here.
The code I linked above is just for getting the logger as you said and using it. It's different for each endpoint (i.e. `http.openapi.operation.healthy` in the example but could be `http.openapi.operation.send` for the endpoint related to the `send` operation and so on).
The two lines we have in the properties file are for assigning a name to the logger related to the specific component in the code. With that name we can configure the logger, in this case just the log level.
I don't think we can get rid of any of them, if this is what you are suggesting and I am getting difficulties to understand :-)

scholzj replied:
So I guess we should maybe replace the lines 14-57 with something like this?

```
Kafka Bridge has its own preconfigured loggers:

* `rootLogger` (default logger for all classes)
* `bridge` (all classes in the `io.strimzi.kafka.bridge` package)
* `healthy` (The healthy endpoint `http.openapi.operation.healthy` - defaults to default: `WARN`)
* `ready` (The healthy endpoint `http.openapi.operation.ready` - defaults to default: `WARN`)

For the pre-configured loggers, you can directly configure the log level.
For example:

[source,yaml,subs="+quotes,attributes"]
----
apiVersion: {KafkaApiVersion}
kind: KafkaBridge
spec:
  # ...
  logging:
    type: inline
    loggers:
      rootLogger.level: INFO
      logger.ready.level: DEBUG
  # ...
----

You can also create custom loggers for the other operations based on HTTP API endpoints:

* `createConsumer`
* `deleteConsumer`
* `subscribe`
* `unsubscribe`
* `poll`
* `assign`
* `commit`
* `send`
* `sendToPartition`
* `seekToBeginning`
* `seekToEnd`
* `seek`
* `openapi`

Each operation has a corresponding API endpoint, defined according to the OpenAPI specification, allowing fine-grained logging of HTTP requests.
Configure each logger by assigning it a name as `http.openapi.operation.<operation_id>`. 
For example, to set the logging level for the send operation logger:

[source,yaml,subs="+quotes,attributes"]
----
apiVersion: {KafkaApiVersion}
kind: KafkaBridge
spec:
  # ...
  logging:
    type: inline
    loggers:
      logger.send.name: "http.openapi.operation.send"
      logger.send.level: DEBUG
  # ...
----
```

ppatierno replied:
Yes I would agree.

scholzj commented on documentation/api/io.strimzi.api.kafka.model.connect.KafkaConnectSpec.adoc:
```diff
@@ -92,19 +92,15 @@ Use a curl request to get a complete list of Kafka Connect loggers running from
-You can set the log levels by specifying the logger and level directly (inline) or use a custom (external) ConfigMap.
-If a ConfigMap is used, you set `logging.valueFrom.configMapKeyRef.name` property to the name of the ConfigMap containing the external logging configuration. Inside the ConfigMap, the logging configuration is described using `log4j.properties`. Both `logging.valueFrom.configMapKeyRef.name` and `logging.valueFrom.configMapKeyRef.key` properties are mandatory. A ConfigMap using the exact logging co ...
-For more information about log levels, see {ApacheLoggers}.
+NOTE: For Strimzi 0.45 and earlier, log4j1 was used for Kafka components. 
+Refer to the Strimzi 0.45 documentation for loggers and logging configuration examples based on log4j1.
```
This is not a question of Strimzi version but Kafka version. So this needs ot be reworked here as well as in the other places. E.g.:

```suggestion
NOTE: Kafka 3.9 and earlier uses Log4j1 got logging. 
Refer to the Strimzi 0.45 documentation for loggers and logging configuration examples based on log4j1 that you can use with Strimzi 0.46 and Kafka 3.9.x: <LINK-TO-THE-DOCS>
```

**We should probably also mark it as IMPORTANT and not NOTE and move it to the top of the section. Ir is pretty lost here in the middle and for users it makes zero sense to start reading this chapter if they use Kafka 3.9 as none of it is valid for them.**

scholzj commented on documentation/api/io.strimzi.api.kafka.model.kafka.KafkaClusterSpec.adoc:
```diff
@@ -154,18 +150,20 @@ spec:
-        log4j.logger.io.strimzi.kafka.oauth: DEBUG
-        log4j.logger.org.openpolicyagents.kafka.OpaAuthorizer: DEBUG
+        rootLogger.level: INFO
+        logger.kafka.name: kafka
```
This line seems to be not needed really?

scholzj commented on documentation/api/io.strimzi.api.kafka.model.kafka.KafkaClusterSpec.adoc:
```diff
@@ -117,32 +117,28 @@ Be aware that if the configured image is not compatible with Strimzi images, it
+Kafka has its own configurable loggers, which include the following:
 
-Kafka uses the Apache `log4j` logger implementation.
+* `rootLogger`
+* `kafka`
+* `orgapachekafka`
+* `requestlogger`
+* `requestchannel`
+* `requestchannel`
+* `controller`
+* `logcleaner`
+* `statechange`
+* `authorizer`
```
Similarly to https://github.com/strimzi/strimzi-kafka-operator/pull/11254#discussion_r2008888734, we should clarify which packages / classes do these loggers cover as it is not completely obvious from their names.

scholzj commented on documentation/api/io.strimzi.api.kafka.model.kafka.KafkaClusterSpec.adoc:
```diff
@@ -154,18 +150,19 @@ spec:
-        log4j.logger.kafka.request.logger: DEBUG
-        log4j.logger.io.strimzi.kafka.oauth: DEBUG
-        log4j.logger.org.openpolicyagents.kafka.OpaAuthorizer: DEBUG
+        rootLogger.level: INFO
+        logger.transaction.name: logger.kafka.coordinator.transaction
+        logger.transaction.level: TRACE
+        logger.cleaner.name: logger.kafka.log.LogCleanerManager
+        logger.cleaner.level: DEBUG
+        logger.request.name: logger.kafka.request.logger
+        logger.request.level: DEBUG
+        logger.oauth.name: logger.io.strimzi.kafka.oauth
+        logger.oauth.level: DEBUG
```
The names should not the loggers. We should also use the predefined loggers here where possible.

```suggestion
        rootLogger.level: INFO
        logger.logcleaner.level: DEBUG
        logger.requestlogger.level: DEBUG
        logger.transaction.name: kafka.coordinator.transaction
        logger.transaction.level: TRACE
        logger.oauth.name: io.strimzi.kafka.oauth
        logger.oauth.level: DEBUG
```

PaulRMellor commented:
Thanks @scholzj 
I've returned to the logger sectiond in the API guide and made the preconfigured logger lists consistent with the change suggested for the bridge.
I've split the examples so the custom loggers are separate.

scholzj reviewed (changes requested):
Sorry, being away from work, I do not have really enough time to focus on the PR all at once. So there is always a bunch of new things I notice with every pass. I imagine it must be hard to keep changing things again and again. Sorry :-(.

PaulRMellor commented on documentation/assemblies/configuring/assembly-logging-configuration.adoc:
```diff
@@ -39,29 +54,32 @@ logging:
-Default logging is used if the `name` or `key` is not set.
+Default logging is used if logging is not specified in the resource using either method.
+Loggers that haven't been explicitly configured inherit settings from their parent loggers. 
+By default, any available loggers that are not configured have their level set to `OFF`.
```
I think it's safe to remove this.

PaulRMellor commented:
Thanks for considering how to make the content clearer @scholzj 
I've separated the inline and external examples, moving the descriptions to introduce each separately

scholzj commented on documentation/modules/configuring/proc-creating-configmap.adoc:
```diff
@@ -6,53 +6,52 @@
 ----
 # Define the logger
-kafka.root.logger.level="INFO"
+rootLogger.level="INFO"
```
We should use here the full properties file as in the previous example:

```suggestion
rootLogger.level = "INFO"
appender.console.type = Console
appender.console.name = STDOUT
appender.console.layout.type = PatternLayout
appender.console.layout.pattern = %d{yyyy-MM-dd HH:mm:ss} %-5p [%t] %c{1}:%L - %m%n
rootLogger.level = INFO
rootLogger.appenderRefs = console
rootLogger.appenderRef.console.ref = STDOUT
rootLogger.additivity = false
```

scholzj commented on documentation/modules/configuring/proc-creating-configmap.adoc:
```diff
@@ -6,53 +6,52 @@
-  log4j.properties:
-    kafka.root.logger.level="INFO"
+  log4j2.properties: |
+    rootLogger.level="INFO"
```
MAybe for the sake of consistency we should use the same spacing

```suggestion
    rootLogger.level = "INFO"
```

scholzj commented on documentation/modules/configuring/proc-creating-logging-filters.adoc:
```diff
@@ -6,59 +6,63 @@
-[source,yaml,subs="+attributes"]
+[source,properties]
 ----
 rootLogger.level="INFO"
```
Same as above about the spacing

```suggestion
rootLogger.level = "INFO"
```

scholzj commented on documentation/modules/configuring/proc-creating-logging-filters.adoc:
```diff
@@ -67,8 +71,16 @@ apiVersion: v1
-  log4j2.properties:
-    #...
+  log4j2.properties: |
+    rootLogger.level="INFO"
```
```suggestion
    rootLogger.level = "INFO"
```

scholzj reviewed (approved):
Few more nits, but LGTM otherwise. I already approved it.
